x/auction/keeper/auction_type.go

```diff
@@ -0,0 +1,103 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/Stride-Labs/stride/v24/x/auction/types"
+)
+
+// Define a type for bid handler functions
+type AuctionBidHandler func(ctx sdk.Context, k Keeper, auction *types.Auction, bid *types.MsgPlaceBid) error
+
+// Map of auction types to their handlers
+var bidHandlers = map[types.AuctionType]AuctionBidHandler{
+	types.AuctionType_AUCTION_TYPE_FCFS: fcfsBidHandler,
+}
+
+// fcfsBidHandler handles bids for First Come First Serve auctions
+func fcfsBidHandler(ctx sdk.Context, k Keeper, auction *types.Auction, bid *types.MsgPlaceBid) error {
+	// Get token amount being auctioned off
+	moduleAddr := k.accountKeeper.GetModuleAddress(types.ModuleName)
```

Comment:
sampocs Dec 17, 2024

should we use a module account per auction? it would make it easier to support multiple auctions with the same denom

assafmo Dec 17, 2024

Yes, but not sure that we want to do it now. I've thought about it a bit, and for the MVP this will complicate the way we send fees to be auctioned off. For each fee token we'll have to map on-chain what auction gets what percentage.

Lmk what you think, I'll do what you decide because I don't have a strong feeling here

Reply:
can you say more on why we would have to figure out percentages? 

The way I was thinking about it was that we would make a module account using the auction name that gets created when you create an auction. And then in stakeibc or when we send fees, we would do something like:
```
for denom in reward collector balance:
    auctionName = f"stakeibc-fees-{denom}"
    auction, found = k.GetAuction(auctionName)
    if not found:
       continue

    k.SendCoins(from: rewardCollector, to: auction.Address)
```

That said, to your point, I don't think it has to be in the MVP, but if it's simple enough like that, I think we should do it

Reply:
My thinking is that it's possible there will be multiple auctions with the same selling token. E.g. maybe in the future we'll want to create 2 auctions for ATOM, one Dutch and the other English, and send 80% of ATOM to the Dutch auction and 20% to the English auction. To implement something like that we can add a `priority` field to auction and then have stakeibc calculate the percentages from the priorities. E.g. priorities of `(1,2,3) translate to percentages `(1/6,2/6,3/6)`.

I think with any of these approaches (all in one account, each auction has an account, each auction has an account + multiple auctions can use same selling denom) we could always migrate to a different approach later on, so I'm not too worried here.

Let's just decide what we want for the MVP and move on.

Reply:
Imo, even in that world of split percentages, it should be the onus of the calling module (i.e. stakeibc) to determine those splits ahead of time - so not something the auction module has to take care of. I also think these two are the same: "each auction has an account" and "each auction has an account + multiple auctions can use same selling denom" cause if you have a different account, the denom doesn't matter.

I'll defer to you here since you're heads more in this / you have a better sense of how much work it would be to split accounts. My mental model with stuff like this is that I find small changes that you know for sure you'll want to make are often better done during the v0 since you're heads most into it and it's a little easier to add it now than, say, 6 months from now where you have to bring context back. 

So if (a) you're aligned this is the best solution, (b) it would be quick to implement and (c) it doesn't increase the testing complexity too much, then I would say probably makes sense to add it. Otherwise, I think punting it is fine. Defer to you!